controller/src/main/java/io/pravega/controller/task/Stream/TxnSweeper.java

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.task.Stream;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.concurrent.FutureHelpers;
+import io.pravega.controller.store.stream.StreamMetadataStore;
+import io.pravega.controller.store.stream.TxnStatus;
+import io.pravega.controller.store.task.TxnResource;
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+
+import java.time.Duration;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * Sweeper for transactions orphaned by failed controller processes.
+ */
+@Slf4j
+public class TxnSweeper {
+    private final StreamMetadataStore streamMetadataStore;
+    private final StreamTransactionMetadataTasks transactionMetadataTasks;
+    private final long maxTxnTimeoutMillis;
+    private final ScheduledExecutorService executor;
+
+    @Data
+    private static class Result {
+        private final TxnResource txnResource;
+        private final Object value;
+        private final Throwable error;
+    }
+
+    public TxnSweeper(final StreamMetadataStore streamMetadataStore,
+                      final StreamTransactionMetadataTasks transactionMetadataTasks,
+                      final long maxTxnTimeoutMillis,
+                      final ScheduledExecutorService executor) {
+        Preconditions.checkNotNull(streamMetadataStore, "streamMetadataStore");
+        Preconditions.checkNotNull(transactionMetadataTasks, "transactionMetadataTasks");
+        Preconditions.checkNotNull(executor, "executor");
+        Preconditions.checkArgument(maxTxnTimeoutMillis > 0, "maxTxnTimeoutMillis should be a positive number");
+
+        this.streamMetadataStore = streamMetadataStore;
+        this.transactionMetadataTasks = transactionMetadataTasks;
+        this.maxTxnTimeoutMillis = maxTxnTimeoutMillis;
+        this.executor = executor;
+    }
+
+    public boolean isReady() {
+        return transactionMetadataTasks.isReady();
+    }
+
+    public void awaitInitialization() throws InterruptedException {
+        transactionMetadataTasks.awaitInitialization();
+    }
+
+    public CompletableFuture<Void> sweepFailedHosts(Supplier<Set<String>> activeHosts) {
+        if (!isReady()) {
+            return FutureHelpers.failedFuture(new IllegalStateException(getClass().getName() + " not yet ready"));
+        }
+        return streamMetadataStore.listHostsOwningTxn().thenComposeAsync(index -> {
```

Comment:
we need to wrap retries around this as we dont want to lose opportunity to sweep a failed host because of an error here.